Distrib: Push the tag only if necessary

Julow opened the pull request:
Check that the tag is already present on the remote and point to the same rev as the local tag.
Don't try to push it again if that's true.

This fixes issue https://github.com/ocamllabs/dune-release/issues/172

Pushing the tag may fail because the guessed URI may be wrong or git may not have access to keys necessary for authentication.
That can happen when the user have aliases for SSH hosts, when the HTTPs protocol is usually used (the user has no key), when the user don't use command line git, or when `dune-release` is sandboxed.

This allows users to push the tag themselves and continue working with the tool when they encounter issues https://github.com/ocamllabs/dune-release/issues/203 and https://github.com/ocamllabs/dune-release/issues/169

Julow commented:
This PR lacks proper testing. I'll try to improve that.

Julow commented:
> it looks like git ls-remote returns tags of the form "refs/tags/ZZZ" wouldn't we need to trim it before testing it against the local tag?
Git is doing that for us and they are not compared. Testing against the local tag is done on the corresponding hash.

`guess_ssh_uri` should be tested indeed. There doesn't seem to be tests for vcs functions setup yet, testing `git_ls_remote` will have to wait for an other PR.

Julow commented on lib/github.ml:
```diff
@@ -50,6 +50,8 @@ module Parse = struct
     | Some _ | None -> error ()
 end
 
+let git_for_repo r = Cmd.of_list (Cmd.to_list @@ Vcs.cmd r)
```
Let's remove that function entirely.

gpetiot commented:
Once #177 has been merged (and fixed first ofc) would it be simpler to just try to push the tag normally and check the error returned by github. There should be an error field in the json response that says something like "this tag already exists". We can consider this is alright instead of failing.
It looks easier to maintain than having a specific request to check that the tag already exists. What do you think?

Julow commented:
It wouldn't. The tag is pushed with git (not API calls) and no errors are returned when it's uptodate.

The problem this PR solves is that pushing the tag requires guessing an URL, which can be wrong.
The goal is to not even try to push the tag when it's not needed.

I think the check implemented in this PR is the easiest to maintain. Git CLI is not going to change soon, unlike Github's API.

Julow commented:
I moved the ls_remote function to the `Vcs` module. This function is only implemented for Git.
For now, it can only be used on Git repositories (it's used when publishing) but this may be surprising if it become useful somewhere else later.
I also rebased and added test cases for the uri guessing function.

mseri commented:
Any update on this. I had issues with tags already presents multiple times.

gpetiot reviewed (approved):
I've rebased and squashed some commits, the last one about the tests of guess_ssh_uri could be extracted in another PR if you want as it does not depend on the rest of this PR. There is not much that can be factorized with #226, but maybe we could have a lib/tag.ml file later on to pack these functions.
It looks good to me, @NathanReb @pitag-ha  what do you think?

gpetiot commented:
@NathanReb @pitag-ha is it all good for you?

pitag-ha reviewed (commented):
Thanks! I think this will make dune-release usable for more people. 

I understand differently what `dev_repo` does (see comment there). I think maybe `ssh_uri_from_http` might not be needed, since the work might already be done by `dev_repo`.

Julow commented on lib/pkg.mli:
```diff
@@ -141,6 +141,7 @@ val extract_tag : t -> (string, Sos.error) result
 (** {1 Dev repo} *)
 
 val dev_repo : t -> (string option, Sos.error) result
+(** dev-repo field with the ["git+"] prefix removed. *)
```
I meant to change the function implementation. Fixed now

pitag-ha replied:
Ah ok. That makes a lot of sense now.

Julow commented on lib/github.ml:
```diff
@@ -235,6 +250,61 @@ let assert_tag_exists ~dry_run tag =
+      Vcs.commit_id ~dirty:false ~commit_ish:tag vcs >>= fun local_rev ->
+      Vcs.ls_remote ~dry_run vcs ~kind:`Tag ~filter:tag dev_repo >>= function
+      | [] -> Ok false
+      | (rev, _) :: _ when local_rev = rev -> Ok true
```
Indeed we need to look for peeled refs (`tag^{}`).

Dune-release normally makes annotated tags but do you think we should do something in case of tags is created by the user and isn't annotated ?
This function don't work in this case right now but perhaps we should catch this before and say it's an error ?

pitag-ha replied:
If we want this to work both for annotated and non-annotated tags, we could look for refs instead of peeled refs (`tag` instead of `tag^{}`) and then compare `local_rev` with `Vcs.commit_id ~dirty:false ~commit_ish:rev vcs`:

```suggestion
      | (rev, _) :: _ when local_rev = Vcs.commit_id ~dirty:false ~commit_ish:rev vcs -> Ok true
```

(if you accept this change, it would be good to also go back to using `tag` instead of `tag^{}`)

Julow replied:
Good idea !
Non-annotated tags still don't work well with the rest of the code (eg. `Pkg.tag` calls `git describe`).
I added a warning message in this case to help diagnostic.

pitag-ha replied:
Thanks.

Julow commented on lib/github.ml:
```diff
@@ -235,6 +250,61 @@ let assert_tag_exists ~dry_run tag =
+        match Parse.ssh_uri_from_http dev_repo with
+        | Some uri -> uri
+        | None ->
+            App_log.unhappy (fun l ->
+                l
+                  "The uri %a is not recognized as a gihub uri, we are going \
+                   to assume it is already a ssh uri."
+                  Text.Pp.url dev_repo);
+            dev_repo
```
This is the only place we interpret the `dev_repo` field, the rest of the function uses the `distrib_uri`.
This is the reason for this PR: The guess is wrong sometimes. This PR adds a way to skip this step by pushing the tag manually.

pitag-ha replied:
I was asking due to the following reason: there are two possibilities:
1. The only way to get into this part of the code is that the guess is wrong. I think in that case it makes sense to report that with `App_log.unhappy` (although in that case we might as well abort directly instead of keeping on going). 
2. It is also possible to get into this part of the code and afterwards the command will still succeed. I was wondering if in that case it might be better to report that with `App_log.status`. What do you think?

I think if dev-repo is of the form ` ssh://git@github.com/...` we might be in possibility 2 (but I'm not sure if people ever use that format, since it's a bit strange). 

Julow replied:
I think it makes sense to keep going. This URI is already meant to be interpreted by Git and that check is only here to work around a common problem: Dev-repo is a Github's HTTP uri but we want to push using Github's SSH uris.
Note that this was already the case and this PR only adds messages and moves code around.

pitag-ha replied:
Ok. I was more talking about the kind of reporting (`unhappy` vs `status`), but I'm not sure myself, so leaving it is good for me.

NathanReb reviewed (commented):
Sorry I had an old review that I forgot to submit somehow so here it is.

I haven't looked at this in a while but please do ping me once those concerns have been addressed and I'll review it right away.

I'm strongly willing to get this merged, once again I appologize for the delay!

NathanReb commented on lib/github.ml:
```diff
@@ -36,6 +36,12 @@ module Parse = struct
         user_from_regexp_opt uri "https://github\\.com/\\(.+\\)/.+\\(\\.git\\)?"
     | _ -> None
+
+  let ssh_uri_from_http uri =
```
cc @gpetiot

NathanReb commented on lib/github.ml:
```diff
@@ -36,6 +36,21 @@ module Parse = struct
+  let ssh_uri_from_http uri =
+    match uri with
+    | _ when Bos_setup.String.is_prefix uri ~affix:"git@" ->
+        path_from_regexp_opt uri "git@github\\.com:\\(.+\\)"
```
Is it really necessary here? we're basically reconstructing the same URI, can't we just return it as it is?

Julow replied:
This is required for something this PR adds:
It prints a warning when guessing the URI is needed but didn't work.

NathanReb commented:
I took a quick look and there are few things that need clarification I think. It seems to me there are two things being addressed here.

The first and most important one is that if the tag is already present on the remote repo, we should not push it. That's the problem most of users get bitten by. This is good behaviour and that part of the PR I most definitely agree we need. Not that this is not github specific at all and therefore should be completely independent from any github dependent feature (even if the rest of the publish command will fail if it's not github) because I think it's important that we clearly distinguish which parts of dune-release rely on github repos and which don't.

There then is the part where we try to infer a URL for pushing. This part I'm not so fond of. The proposed behaviour is to always push using SSH, infering the URL from the dev-repo. We already know some users don't use SSH for pushing and that this behaviour is a problem for them. @Julow am I write that the issue you have is that you use custom URL schemes for your gh repos and that doesn't work well with dune-release atm? If we fix the tag pushing behaviour mentioned above and that you push the tag ahead of running `dune-release publish`, would that fix it for you or will dune-release fail to fetch the tags? E.g. if we fetch the tags using http and the dev-repo, we don't need to figure how to push if the tag's already there.

Julow commented:
> There then is the part where we try to infer a URL for pushing. This part I'm not so fond of.

No, this PR is not adding that, this URI was already being "converted" to ssh URIs before. This PR moves code around to make things clearer and changes the behavior a bit: Print a warning if the dev_repo is not a Github URI.

This PR doesn't solve the problem of guessing the URIs, it is adding a way to avoid it.

About my problem, yes, this PR allows me to push the tag manually and then run `distrib`. The guessed URI is now only used for pushing the tag, not for fetching it (the original `dev_repo` is used for that). And yes, that's how this PR works, the URI doesn't need to be guessed at all if the tag was already pushed.

Julow commented on lib/github.ml:
```diff
@@ -226,39 +230,74 @@ let assert_tag_exists ~dry_run tag =
+          Ok false
+  in
+  remote_has_tag_uptodate () >>= function
+  | true -> Ok () (* No need to push, avoiding the need to guess the uri. *)
```
Good idea. I added that.

Julow commented on lib/github.ml:
```diff
@@ -226,39 +230,74 @@ let assert_tag_exists ~dry_run tag =
+      | (rev, _) :: _ ->
+          App_log.unhappy (fun l ->
+              l
+                "The tag %a is present on the remote but point to a different \
```
It seems to be already fixed.

NathanReb commented:
Ah my bad, thanks for clarifying all this! It's seems good then!

Hopefully we'll soon improve the URL management in such a way that you'll be able to use your own config. Since we're having troubles with HTTP vs SSH for write access, I was thinking about adding a config entry for this. If we go down that road we could allow:
- `http` which corresponds to using the https remote github url
- `ssh` which corresponds to using the `git@github.com:user/repo` github specific ssh scheme
- `remote:<remote name>` which would push to the given remote name, be it origin, upstream or whatever one wants to use.

What would you think about this?

Julow commented on lib/github.ml:
```diff
@@ -226,39 +230,74 @@ let assert_tag_exists ~dry_run tag =
+let push_tag ~dry_run ~yes ~dev_repo vcs tag =
+  let git = Vcs.cmd vcs in
+  let remote_has_tag_uptodate () =
+    if dry_run then Ok true (* Assume yes when [dry_run] is enabled *)
```
Indeed, I couldn't find a reason. Changed that.

Julow commented:
A config option would be useful. If we remove the code guessing uris, it could look like this:
- nothing, use the `dev_repo` without interpreting it
- something, will use that without interpreting it. Could be a remote name, a http or a ssh URI, Git will know.

NathanReb commented:
It won't work like this because I don't want to introduce local configuration files for now so it would be a global config thing.

Also I don't want to change the default behaviour (push to SSH) as it already works for most users, using `dev-repo` means switching to http in most cases and therefore would break everyone's workflow.

Julow commented:
I'm not sure a global config file could help here. What if two projects require different configs while an other works fine with the default ?
The current behavior seems very adhoc and adding complexity to it would make things very hard to understand and fix for a user.
The wrong part is guessing URIs, maybe we should try to remove this completely ?

What about an interactive prompt ? Here are the URIs to choose from:
- `dev-repo`
- `git remote -v`
- the guessed SSH URI for Github
- let the user type an URI.

That has the advantage of not breaking existing workflows and that the guessed URI is still here for users that need it.

In my specific case, the URI that would be able to push a tag to the upstream repository is not any of these and I'd need to enter it.

NathanReb commented on lib/github.ml:
```diff
@@ -235,6 +250,79 @@ let assert_tag_exists ~dry_run tag =
+  remote_has_tag_uptodate () >>= function
+  | true ->
+      App_log.status (fun l ->
+          l "The tag %a is present and uptodate on the remote." Text.Pp.version
```
Sorry to be so picky but could you change the message to:
```suggestion
          l "The tag %a is present and uptodate on the remote: skipping the tag push" Text.Pp.version
```
To clearly state what is going on here, otherwise it's unclear that anything is skipped from the regular workflow and new user might get confused. They might think it's some kind of check and that it's their responsibility to push the tag for instance.

pitag-ha commented on lib/github.ml:
```diff
@@ -235,6 +250,79 @@ let assert_tag_exists ~dry_run tag =
+                  "The tag %a is present on the remote but points to a \
+                   different commit (%a)."
+                  Text.Pp.version tag pp_r r);
+            Ok false )
```
I would have expected an `Error` here. Do you want to try pushing the tag even though the tag already exists on the remote (pointing to a wrong commit)? Does github allow that?

Julow replied:
The push has `--force`.
No need to fail here, the output will be this warning followed by the question "Push tag ... ?", which would work as usual.

pitag-ha replied:
Ok, perfect then.

pitag-ha commented on lib/github.ml:
```diff
@@ -235,6 +250,79 @@ let assert_tag_exists ~dry_run tag =
+    Vcs.ls_remote ~dry_run vcs ~kind:`Tag ~filter:tag dev_repo >>= function
+    | [] -> Ok false
+    | (remote_rev_unpeeled, _) :: _ -> (
+        (* Resolve again in case of annotated commits (most common case).
+           This is a no-op for non-annotated commits. In case of error, we
```
```suggestion
        (* Resolve again in case of annotated tags (most common case).
           This is a no-op for non-annotated tags. In case of error, we
```

Julow replied:
Multiline suggestion !

pitag-ha reviewed (commented):
Thanks for the changes! I just have one more question.

Julow commented:
Just rebased to have the CI fixed.

NathanReb commented:
Ok let's merge this!
